src/net/connection/connection_pool.rs

```diff
@@ -26,16 +26,23 @@ impl ConnectionPool {
 
     /// Try getting connection by address if the connection does not exists it will be inserted.
     pub fn get_connection_or_insert(&self, addr: &SocketAddr) -> NetworkResult<Connection> {
-        let mut lock = self
-            .connections
-            .write()
+        let lock =  self.connections
+            .read()
             .map_err(|error| NetworkError::poisoned_connection_error(error.description()))?;
 
-        let connection = lock
-            .entry(*addr)
-            .or_insert_with(|| Arc::new(RwLock::new(VirtualConnection::new(*addr, &self.config))));
+        if lock.contains_key(addr) {
+            return Ok(lock.get(addr).unwrap().clone());
```

Comment:
Can you remove the return and the unwrap?

Reply:
Strange though I already did that